Group Bookings DRT

nkuehnel opened the pull request.

nkuehnel commented on contribs/drt/src/main/java/org/matsim/contrib/drt/passenger/DrtStopActivity.java:
```diff
@@ -67,39 +69,39 @@ protected boolean isLastStep(double now) {
 		return pickupRequests.values()
 				.stream()
-				.filter(r -> passengerId.equals(r.getPassengerId()))
+				.filter(r -> r.getPassengerIds().containsAll(passengerIds))
```
make sure all passengers are part of the request

nkuehnel commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/AbstractPassengerRequestEvent.java:
```diff
@@ -35,19 +37,20 @@
  * @author Michal Maciejewski (michalm)
  */
-public abstract class AbstractPassengerRequestEvent extends Event implements HasPersonId {
+public abstract class AbstractPassengerRequestEvent extends Event {
```
@michalmac since all AbstractPassengerRequestEvents can now be about multiple persons (submitted, scheduled, rejected) we can no longer implement the HasPersonId interface which would only return one person id
@michalmac

nkuehnel commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/AbstractPassengerRequestEvent.java:
```diff
@@ -62,18 +65,18 @@ public final Id<Request> getRequestId() {
 		Map<String, String> attr = super.getAttributes();
 		attr.put(ATTRIBUTE_MODE, mode);
 		attr.put(ATTRIBUTE_REQUEST, requestId + "");
+		attr.put(ATTRIBUTE_PERSON, personIds.stream().map(Object::toString).collect(Collectors.joining(",")));
```
here I'm joining the ids with a "," so far...

nkuehnel commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/InternalPassengerHandling.java:
```diff
@@ -70,39 +70,50 @@ boolean validateRequest(PassengerRequest request, PassengerRequestValidator requ
-			//only possible with prebooking
-			return false;
+
+		//ensure for every passenger first
```
proceed only if all passengers are present

nkuehnel commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/InternalPassengerHandling.java:
```diff
@@ -70,39 +70,50 @@ boolean validateRequest(PassengerRequest request, PassengerRequestValidator requ
+
+			Id<DvrpVehicle> vehicleId = Id.create(mobVehicle.getId(), DvrpVehicle.class);
+			eventsManager.processEvent(new PersonEntersVehicleEvent(now, passenger.getId(), mobVehicle.getId()));
+			eventsManager.processEvent(new PassengerPickedUpEvent(now, mode, requestId, passenger.getId(), vehicleId));
```
process each agent individually and throw pickup event separately

nkuehnel commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/InternalPassengerHandling.java:
```diff
@@ -70,39 +70,50 @@ boolean validateRequest(PassengerRequest request, PassengerRequestValidator requ
+			mobVehicle.removePassenger(passenger);
+			passenger.setVehicle(null);
+
+			eventsManager.processEvent(new PassengerDroppedOffEvent(now, mode, requestId, passenger.getId(), vehicleId));
```
separate dropoff events per passenger as well

nkuehnel commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/MultiPassengerPickupActivity.java:
```diff
@@ -49,32 +52,32 @@ public MultiPassengerPickupActivity(PassengerHandler passengerHandler, DynAgent
-			if (passengerHandler.tryPickUpPassenger(this, driver, request.getId(), now)) {
-				passengersPickedUp++;
+			if (passengerHandler.tryPickUpPassengers(this, driver, request.getId(), now)) {
+				requestsPickedUp++;
```
count complete request pickups instead of passengers

nkuehnel commented:
So far this seems to work with the existing tests and implementations for the single-passenger case.

A few notes:
- AbstractPassengerRequest events are now handled _once_ per request and may refer to multiple passengers (submitted, scheduled, rejected events)
- AbstractPassengerEvents are so far still thrown _per_ passenger of the request (departure, pickup, dropoff events)

In the analysis part, each passenger will still be recorded as a single DrtLeg so far.

michalmac commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/PassengerRequestSubmittedEvent.java:
```diff
@@ -41,9 +45,9 @@ public class PassengerRequestSubmittedEvent extends AbstractPassengerRequestEven
 	private final Id<Link> toLinkId;
 
-	public PassengerRequestSubmittedEvent(double time, String mode, Id<Request> requestId, Id<Person> personId,
+	public PassengerRequestSubmittedEvent(double time, String mode, Id<Request> requestId, Set<Id<Person>> personIds,
```
Please change from `Set` to `List` in all events. `List` is better for events (computations + memory).

BTW. Using a list does not mean that we cannot guarantee uniqueness of elements.

michalmac commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/SinglePassengerPickupActivity.java:
```diff
@@ -44,22 +48,22 @@ public SinglePassengerPickupActivity(PassengerHandler passengerHandler, DynAgent
-	public void notifyPassengerIsReadyForDeparture(MobsimPassengerAgent passenger, double now) {
-		if (passenger.getId().equals(request.getPassengerId())) {
+	public void notifyPassengersAreReadyForDeparture(Set<MobsimPassengerAgent> passengers, double now) {
+		if (request.getPassengerIds().containsAll(passengers.stream().map(Identifiable::getId).collect(Collectors.toSet()))) {
```
`passengers.stream().map(Identifiable::getId).collect(Collectors.toSet())` is unnecessarily heavy. We can simply collect to list:

`passengers.stream().map(Identifiable::getId).toList()`

michalmac commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/TeleportingPassengerEngine.java:
```diff
@@ -143,10 +143,10 @@ public boolean handleDeparture(double now, MobsimAgent agent, Id<Link> fromLinkI
 		Route route = ((Leg)((PlanAgent)passenger).getCurrentPlanElement()).getRoute();
 		PassengerRequest request = requestCreator.createRequest(internalPassengerHandling.createRequestId(),
-				passenger.getId(), route, getLink(fromLinkId), getLink(toLinkId), now, now);
+				Collections.singleton(passenger.getId()), route, getLink(fromLinkId), getLink(toLinkId), now, now);
```
Use `List.of()` or `Set.of()` instead of their counterparts in `Collections`

michalmac commented on contribs/dvrp/src/test/java/org/matsim/contrib/dvrp/examples/onetaxi/RunOneTaxiWithPrebookingExampleIT.java:
```diff
@@ -239,7 +239,7 @@ private static void assertRequestScheduledEvent(Map<Id<Person>, PassengerRequest
 		assertThat(event.getVehicleId().toString()).isEqualTo("taxi_one");
 		assertThat(event.getPickupTime()).isCloseTo(pickupTime, Offset.offset(0.01));
-		assertThat(event.getPersonId().toString()).isEqualTo(personId);
+		assertThat(event.getPersonIds().stream().findAny().orElseThrow().toString()).isEqualTo(personId);
```
I would rather assert that size is 1 and then take the first element (BTW. after changing to a list, it will be easier to take the first element).

michalmac commented on contribs/dvrp/src/test/java/org/matsim/contrib/dvrp/passenger/PassengerEngineTestFixture.java:
```diff
@@ -107,7 +107,10 @@ void addPersonWithLeg(Link fromLink, Link toLink, double departureTime) {
-				.filter(e -> e instanceof HasPersonId && ((HasPersonId)e).getPersonId().equals(PERSON_ID));
+				.filter(e ->
+						e instanceof HasPersonId && ((HasPersonId)e).getPersonId().equals(PERSON_ID) ||
+								e instanceof AbstractPassengerRequestEvent
```
Why is this change made?

nkuehnel replied:
@michalmac because the abstractpassengerrequest events do not implement the HasPersonid anymore, as they do not point to a single person any longer

nkuehnel commented:
@michalmac thanks for the review. I changed the sets to lists in the respective places. 
There's also a GroupPassengerEngine implementation now that can still be moved to a separate PR if required. I had to duplicate the DefaultPassengerEngne class, as delegation would not have worked as there are some internal fields (e.g.,  activePassengers) that can not be properly handled by simple delegation.

It's ready for re-review from my point of view :)

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/analysis/DrtEventSequenceCollector.java:
```diff
@@ -75,27 +75,24 @@ public static class EventSequence {
-		private PassengerDroppedOffEvent droppedOff;
+		private Map<Id<Person>, PersonDepartureEvent> departures = new HashMap<>();
+		private Map<Id<Person>, PassengerPickedUpEvent> pickedUp = new HashMap<>();
+		private Map<Id<Person>, PassengerDroppedOffEvent> droppedOff = new HashMap<>();
```
Have you thought about having a record `PersonEvents` with these 3 fields: `PersonDepartureEvent`, `PassengerPickedUpEvent` and `PassengerDroppedOffEvent`? Then we would only need one map: `Map<Id<Person>, PersonEvents>`. That should be less heavy for bigger simulations.

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/analysis/zonal/DrtZonalWaitTimesAnalyzer.java:
```diff
@@ -138,11 +140,13 @@ private Map<String, DescriptiveStatistics> createZonalStats() {
-				final String zoneStr = zone != null ? zone.getId() : zoneIdForOutsideOfZonalSystem;
-				double waitTime = seq.getPickedUp().get().getTime() - seq.getSubmitted().getTime();
-				zoneStats.get(zoneStr).addValue(waitTime);
+			for (Id<Person> personId : seq.getSubmitted().getPersonIds()) {
+				if(seq.getPickedUp().containsKey(personId)) {
```
It's enough to simply loop through `seq.getPickedUp()` entries.

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionGenerator.java:
```diff
@@ -156,7 +156,7 @@ public List<InsertionWithDetourData> generateInsertions(DrtRequest drtRequest, V
 			Waypoint.Stop nextStop = nextStop(vEntry, i);
 
-			if (occupancy < vEntry.vehicle.getCapacity()) {// only not fully loaded arcs
+			if (occupancy + drtRequest.getPassengerIds().size() <= vEntry.vehicle.getCapacity()) {// only not fully loaded arcs
```
Maybe it would be nice to have `DrtRequest.size()` method?

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionGenerator.java:
```diff
@@ -222,7 +222,7 @@ private void generateDropoffInsertions(DrtRequest request, VehicleEntry vEntry,
 			// check the capacity constraints if i < j (already validated for `i == j`)
 			Waypoint.Stop currentStop = currentStop(vEntry, j);
-			if (currentStop.outgoingOccupancy == vEntry.vehicle.getCapacity()) {
+			if (vEntry.vehicle.getCapacity() - currentStop.outgoingOccupancy < request.getPassengerIds().size()) {
```
Maybe this is easier to read (and more aligned with the change in L166):
`currentStop.outgoingOccupancy + request.getPassengerIds().size() > vEntry.vehicle.getCapacity()`

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/passenger/DrtRequest.java:
```diff
@@ -109,8 +111,8 @@ public Link getToLink() {
-	public Id<Person> getPassengerId() {
-		return passengerId;
+	public List<Id<Person>> getPassengerIds() {
+		return Collections.unmodifiableList(passengerIds);
```
In this and other similar places, consider using the "more modern" Collections API, e.g. `List.copyOf()` (the created list is also unmodifiable).

michalmac commented on contribs/drt-extensions/src/main/java/org/matsim/contrib/drt/extension/operations/eshifts/charging/ChargingBreakActivity.java:
```diff
@@ -95,9 +98,9 @@ protected void simStep(double now) {
+    private AcceptedDrtRequest getRequestForPassengers(Set<Id<Person>> passengerIds) {
         return pickupRequests.values().stream()
-                .filter(r -> passengerId.equals(r.getPassengerId()))
+                .filter(r -> r.getPassengerIds().containsAll(passengerIds))
```
This will ensure the two lists contain exactly the same elements (assuming there are no duplicates) and it should be quicker to discard the requests if many of them have different sizes.
```java
r.getPassengerIds() == passengerIds.size() && r.getPassengerIds().containsAll(passengerIds)
```

michalmac commented on contribs/drt-extensions/src/main/java/org/matsim/contrib/drt/extension/preplanned/optimizer/PreplannedDrtOptimizer.java:
```diff
@@ -203,7 +201,7 @@ public record PreplannedSchedules(Map<PreplannedRequestKey, Id<DvrpVehicle>> pre
 	}
 
-	public record PreplannedRequestKey(Id<Person> passengerId, Id<Link> fromLinkId, Id<Link> toLinkId) {
+	public record PreplannedRequestKey(List<Id<Person>> passengerIds, Id<Link> fromLinkId, Id<Link> toLinkId) {
```
Here, it may be better to have a set instead of a list, because two "seemingly identical" keys, but with different order of passengerIds will not be identical.

nkuehnel commented:
@michalmac I addressed your comments. Since quite a few things have changed after the prebooking feature I would appreciate another quick review @sebhoerl @michalmac  :) 

Otherwise I would be ready to merge

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/analysis/DrtAnalysisControlerListener.java:
```diff
@@ -162,24 +169,25 @@ public void notifyIterationEnds(IterationEndsEvent event) {
 					Coord toCoord = network.getLinks().get(submission.getToLinkId()).getToNode().getCoord();
 					return String.join(delimiter, submission.getTime() + "",//
-							submission.getPersonId() + "",//
+							submission.getPersonIds().stream().map(Object::toString).collect(Collectors.joining("-")) + "",//
```
Sounds good.

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/analysis/DrtEventSequenceCollector.java:
```diff
@@ -68,31 +68,50 @@ public class DrtEventSequenceCollector
 
 	public static class EventSequence {
+
+		public final static class PersonEvents {
```
Just wondering why not a record?

nkuehnel replied:
@michalmac because fields in records are final and events are not added sequentially to the PersonEvents object

michalmac commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/PassengerRequest.java:
```diff
@@ -46,4 +46,6 @@ default double getLatestStartTime() {
 
 	String getMode();
+
+	int getPartySize();
```
`getPartySize()` may be a confusing to others. IMO `getSize()` or `getPassengerCount()` are more self-descriptive.

If you want to use `getPartySize()`, then I suggest adding a javadoc comment.

nkuehnel replied:
Thanks, I'll go with PassengerCount

michalmac commented on contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/GroupPassengerEngine.java:
```diff
@@ -0,0 +1,231 @@
+			// must come before validateAndSubmitRequest (to come before rejection event)
+			eventsManager.processEvent(new PassengerWaitingEvent(now, mode, request.getId(), group.stream().map(Identifiable::getId).toList()));
+
+			validateAndSubmitRequest(List.copyOf(group), request, mobsimTimer.getTimeOfDay());
```
Nit: No need to create an immutable copy when passing the list to a private method (inside your class).

michalmac commented:
> @michalmac I addressed your comments. Since quite a few things have changed after the prebooking feature I would appreciate another quick review @sebhoerl @michalmac :)
> 
> Otherwise I would be ready to merge

I quicky went through the code changes. I think it is ready to get merged.
